wfepy/workflow.py

```diff
@@ -242,6 +246,7 @@ def _step(self, state):
                 elif result:
                     logger.info('Task %s is complete', task_name)
                     next_state.append((task_name, TaskState.COMPLETE))
+                    self.completed_tasks.append(task_name)
```

Comment:
Modifying `self.completed_tasks` in this method does not seem correct because it will be out-of-sync with `self.state` potentially.

Reply:
Would it be possible to keep COMPLETE in state instead and later see if we can add NEW tasks based on this?

Reply:
Close this patch and think about to re-implement the whole looping logic